src/ameba/rule/style/verbose_block.cr

```diff
@@ -0,0 +1,229 @@
+      String.build do |str|
+        case name = node.name
+        when "[]"
+          str << '['
+          args_to_s(str, node)
+          str << ']'
+        when "[]?"
+          str << '['
+          args_to_s(str, node)
+          str << "]?"
+        when "[]="
+          str << '['
+          args_to_s(str, node, skip_last_arg: true)
+          str << "]=(" << node.args.last? << ')'
+        else
+          str << name
+          if !node.args.empty? || (node.named_args && !node.named_args.try(&.empty?))
+            str << '('
+            args_to_s(str, node)
+            str << ')'
+          end
+          str << " {...}" if node.block
+        end
+      end
+    end
+
+    protected def call_code(call, body)
+      args = call.args.join ", " unless call.args.empty?
+      args += ", " if args
+
+      call_chain = %w[].tap do |arr|
+        obj = body.obj
+        while obj.is_a?(Crystal::Call)
+          arr << node_to_s(obj)
+          obj = obj.obj
+        end
+        arr.reverse!
+        arr << node_to_s(body)
+      end
+
+      name =
+        call_chain.join('.')
+
+      CALL_PATTERN % {call.name, args, name}
+    end
+
+    # ameba:disable Metrics/CyclomaticComplexity
+    protected def issue_for_valid(source, call : Crystal::Call, body : Crystal::Call)
+      return if exclude_calls_with_block && body.block
+      return if exclude_multiple_line_blocks && !same_location_lines?(call, body)
+      return if exclude_prefix_operators && prefix_operator?(body)
+      return if exclude_operators && operator?(body.name)
+      return if exclude_setters && setter?(body.name)
+
+      call_code =
+        call_code(call, body)
+
+      return unless valid_line_length?(call, call_code)
+      return unless valid_length?(call_code)
+
+      issue_for call.name_location, call.end_location,
+        MSG % call_code
+    end
+
+    def test(source, node : Crystal::Call)
+      # we are interested only in calls with block taking a single argument
+      #
+      # ```
+      # (1..3).any? { |i| i.to_i64.odd? }
+      #        ^---    ^  ^------------
+      #        block  arg  body
+      # ```
+      return unless (block = node.block) && block.args.size == 1
+
+      arg = block.args.first
+
+      # we skip auto-generated blocks - `(1..3).any?(&.odd?)`
+      return if arg.name.starts_with?("__arg")
+
+      # we filter out the blocks that are of call type - `i.to_i64.odd?`
+      return unless (body = block.body).is_a?(Crystal::Call)
+
+      # we need to "unwind" the chain challs, so the final receiver object
+      # ends up being a variable - `i`
+      obj = body.obj
+      while obj.is_a?(Crystal::Call)
+        obj = obj.obj
+      end
+      return unless obj.is_a?(Crystal::Var)
+
+      # only calls with a first argument used as a receiver are the valid game
+      return unless obj == arg
+
+      # we bail out if the block node include the block argument
+      return if reference_count(body, arg) > 1
```

Comment:
I think this can be a bit optimized. Since we don't need to count all the references, we just need to return if at least two are found.

Reply:
That's correct. Would you like it to be done in this PR or could it wait for the followup?

Reply:
This is not very important. I’m totally fine with a followup. Thanks, @Sija